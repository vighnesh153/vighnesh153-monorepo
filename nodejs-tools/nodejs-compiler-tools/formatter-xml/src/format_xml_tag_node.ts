import { XmlTagNode, XmlTextNode } from '@vighnesh153/parser-xml';
import { buildIndentationSpace } from './build_indentation_space';
import { formatXmlElementAttribute } from './format_xml_element_attribute';
import { formatXmlExpression } from './format_xml_expression';
import { sortAttributes } from './sort_attributes';
import { formatTextNode } from './format_text_node';

type FormatXmlTagNodeConfig = {
  xmlTagNode: XmlTagNode;
  indentationLevel: number;
  indentation: number;
  sortAttributes: boolean;
};

export function formatXmlTagNode({
  xmlTagNode,
  indentationLevel,
  indentation,
  sortAttributes: shouldSortAttributes,
}: FormatXmlTagNodeConfig): string {
  const { namespaces, attributes, children } = xmlTagNode;

  const tagName = namespaces.map((ns) => ns.tokenLiteral).join(':');

  const stringBuilder: string[] = [];

  // push: <tag-name
  stringBuilder.push(`${buildIndentationSpace({ indentationLevel, indentation })}<${tagName}`);

  // push attributes
  if (attributes.length === 1) {
    stringBuilder[stringBuilder.length - 1] += ' ' + formatXmlElementAttribute(attributes[0]);
  } else if (attributes.length > 1) {
    (shouldSortAttributes ? sortAttributes(attributes) : attributes)
      .map((attribute) => formatXmlElementAttribute(attribute))
      .forEach((formattedAttribute) => {
        stringBuilder.push(
          `${buildIndentationSpace({ indentation, indentationLevel: indentationLevel + 1 })}${formattedAttribute}`
        );
      });
  }

  if (children.length === 0) {
    // push self closing tag's end marker
    stringBuilder[stringBuilder.length - 1] = stringBuilder.at(-1) + ' />';
  } else {
    // push tag opening's end marker
    stringBuilder[stringBuilder.length - 1] = stringBuilder.at(-1) + '>';

    // if only a text node as child, format it as
    if (children.length === 1 && children[0].astNodeType === 'XML_TEXT_NODE') {
      return (
        stringBuilder.join('\n') +
        formatTextNode({ textNode: children[0] as XmlTextNode, indentation, indentationLevel: 0 }) +
        `</${tagName}>`
      );
    }

    // push children
    for (const child of children) {
      stringBuilder.push(
        formatXmlExpression({
          expression: child,
          indentation,
          indentationLevel: indentationLevel + 1,
          sortAttributes: shouldSortAttributes,
        })
      );
    }

    // push closing tag
    stringBuilder.push(`${buildIndentationSpace({ indentationLevel, indentation })}</${tagName}>`);
  }

  return stringBuilder.join('\n');
}
