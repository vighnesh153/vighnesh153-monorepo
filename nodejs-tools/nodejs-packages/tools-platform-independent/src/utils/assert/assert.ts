import { not } from '../not';

export function assert(condition: boolean, message = 'Assertion failed') {
  if (not(condition)) {
    throw new Error(message);
  }
}
