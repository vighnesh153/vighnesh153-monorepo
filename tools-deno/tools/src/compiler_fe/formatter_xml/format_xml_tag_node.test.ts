import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import type { XmlTagNode } from "@/compiler_fe/parser_xml/mod.ts";
import { formatXmlTagNode } from "./format_xml_tag_node.ts";
import { parseProgram } from "./test_utils.ts";

Deno.test("should format empty, self-closing xml tag node", async (t) => {
  const [parser, program] = parseProgram(`<manifest />`);

  assertEquals(parser.errors.length, 0);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(
    t,
    formatXmlTagNode({
      xmlTagNode: program.statements[0] as XmlTagNode,
      indentation: 4,
      indentationLevel: 0,
      sortAttributes: true,
    }),
  );
});

Deno.test("should format self-closing xml tag node with attributes", async (t) => {
  const [parser, program] = parseProgram(
    `<manifest   simpleAttribute = "some random value" deeply:nested:attribute = "20" />`,
  );

  assertEquals(parser.errors.length, 0);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(
    t,
    formatXmlTagNode({
      xmlTagNode: program.statements[0] as XmlTagNode,
      indentation: 4,
      indentationLevel: 0,
      sortAttributes: true,
    }),
  );
});

Deno.test("should format non self-closing xml tag node with attributes but no children", async (t) => {
  const [parser, program] = parseProgram(
    `<manifest   simpleAttribute = "some random value" deeply:nested:attribute = "20" >
  </ manifest>
    `,
  );

  assertEquals(parser.errors.length, 0);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(
    t,
    formatXmlTagNode({
      xmlTagNode: program.statements[0] as XmlTagNode,
      indentation: 4,
      indentationLevel: 0,
      sortAttributes: true,
    }),
  );
});

Deno.test("should format xml tag node with children", async (t) => {
  const [parser, program] = parseProgram(
    `<manifest    >
     < childTag  special:forces:commando = "Pikachu"  pokemon= "Hurray" />
      < childTag2    pokemon= "Greninja">
          <  childTag3  Infernape = "lol" >  </ childTag3>
       </ childTag2>
  </ manifest>
    `,
  );

  assertEquals(parser.errors.length, 0);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(
    t,
    formatXmlTagNode({
      xmlTagNode: program.statements[0] as XmlTagNode,
      indentation: 4,
      indentationLevel: 0,
      sortAttributes: true,
    }),
  );
});

Deno.test("should sort all attributes based on android rules", async (t) => {
  const [parser, program] = parseProgram(
    `<manifest  
        qlaAttr="prop18"
        name="prop14"
        blaAttr="prop16"
        xmlns:android="prop1"
        android:bla:id="prop5"
        android:zza:name="prop12"
        android:plz:name="prop11"
        android:poki2:id="prop8"
        android:raster="prop22"
        xmlns:app="prop2"
        android:pae="prop21"
        zools:duckie="prop27"
        plaAttr="prop17"
        tools:hungry="prop26"
        style="prop15"
        android:zzb:name="prop13"
        android:id="prop4"
        android:plu:name="prop10"
        zlaAttr="prop19"
        app:bla="prop24"
        xmlns:tools="prop3"
        android:blb:id="prop6"
        android:name="prop9"
        android:layout="prop20"
        app:pla="prop25"
        android:poki1:id="prop7"
        android:temper="prop23"
     />
    `,
  );

  assertEquals(parser.errors.length, 0);
  assertEquals(program.statements.length, 1);
  await assertSnapshot(
    t,
    formatXmlTagNode({
      xmlTagNode: program.statements[0] as XmlTagNode,
      indentation: 4,
      indentationLevel: 0,
      sortAttributes: true,
    }),
  );
});

Deno.test(
  "should format xml tag node without sorting if sortAttribute is false",
  async (t) => {
    const [parser, program] = parseProgram(
      `<manifest    >
     < childTag  special:forces:commando = "Pikachu"  pokemon= "Hurray" />
      < childTag2    pokemon= "Greninja">
          <  childTag3  Infernape = "lol" >  </ childTag3>
       </ childTag2>
  </ manifest>
    `,
    );

    assertEquals(parser.errors.length, 0);
    assertEquals(program.statements.length, 1);
    await assertSnapshot(
      t,
      formatXmlTagNode({
        xmlTagNode: program.statements[0] as XmlTagNode,
        indentation: 4,
        indentationLevel: 0,
        sortAttributes: false,
      }),
    );
  },
);
