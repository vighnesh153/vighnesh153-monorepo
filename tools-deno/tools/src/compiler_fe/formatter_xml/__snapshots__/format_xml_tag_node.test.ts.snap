export const snapshot = {};

snapshot[`should format empty, self-closing xml tag node 1`] = `"<manifest />"`;

snapshot[`should format self-closing xml tag node with attributes 1`] = `
'<manifest
    simpleAttribute="some random value"
    deeply:nested:attribute="20" />'
`;

snapshot[`should format non self-closing xml tag node with attributes but no children 1`] = `
'<manifest
    simpleAttribute="some random value"
    deeply:nested:attribute="20" />'
`;

snapshot[`should format xml tag node with children 1`] = `
'<manifest>
    <childTag
        pokemon="Hurray"
        special:forces:commando="Pikachu" />
    <childTag2 pokemon="Greninja">
        <childTag3 Infernape="lol" />
    </childTag2>
</manifest>'
`;

snapshot[`should sort all attributes based on android rules 1`] = `
'<manifest
    xmlns:android="prop1"
    xmlns:app="prop2"
    xmlns:tools="prop3"
    android:id="prop4"
    android:bla:id="prop5"
    android:blb:id="prop6"
    android:poki1:id="prop7"
    android:poki2:id="prop8"
    android:name="prop9"
    android:plu:name="prop10"
    android:plz:name="prop11"
    android:zza:name="prop12"
    android:zzb:name="prop13"
    name="prop14"
    style="prop15"
    blaAttr="prop16"
    plaAttr="prop17"
    qlaAttr="prop18"
    zlaAttr="prop19"
    android:layout="prop20"
    android:pae="prop21"
    android:raster="prop22"
    android:temper="prop23"
    app:bla="prop24"
    app:pla="prop25"
    tools:hungry="prop26"
    zools:duckie="prop27" />'
`;

snapshot[`should format xml tag node without sorting if sortAttribute is false 1`] = `
'<manifest>
    <childTag
        special:forces:commando="Pikachu"
        pokemon="Hurray" />
    <childTag2 pokemon="Greninja">
        <childTag3 Infernape="lol" />
    </childTag2>
</manifest>'
`;
